Rep 2233 fix

dkowis opened the pull request:
This fixes REP-2233 by enforcing via XSD validation that all limit IDs across all limit-groups are unique.

EDIT: I think this is only a partial fix. I'm going to write another test to verify. I think the bug will resurface if multiple `rate-limiting-filter`s are used, there could be conflicting IDs in those files (since the logic happens in the RateLimitingService, not the filter.

EDIT: It's not a partial fix, I added another test to it to prove that it behaves well. Yay.

dkowis commented:
Derp I forgot to reformat this :( getting narc'ed on.

jennyvo commented:
Retest this please.

jennyvo commented:
4 tests failed:
- SchemaTests.java - fixed
- IntraFilterLoggingTest - don't know why - didn't log last 4 filters after the rate-limiting filter 
- DatastoreWarnLimitTest - don't know why
- PatchMethodTest - don't know why - timeout

jennyvo commented:
Fix all the tests failed --all because rate-limiting config violate unique id.

jennyvo commented:
Retest this please.

dkowis commented:
retest this please

dkowis commented:
Something seems to be buggered with Jenkins :(

jennyvo commented:
yep, other pull request also didn't kick of run from jenkin.

wdschei commented:
Retest this please.

wdschei commented on repose-aggregator/services/rate-limiting-service/src/main/resources/META-INF/schema/config/rate-limiting-configuration.xsd:
```diff
@@ -162,6 +162,10 @@
+                   xerces:message="ALL Limits must have unique ids across all limit groups"
+                   saxon:message="ALL Limits must have unique ids across all limit groups" xpathDefaultNamespace="##targetNamespace"/>
+
         <!-- TODO: assert at least one limit-group or global-limit-group is present -->
```
Easy enough.

dkowis replied:
The to-do was there before I added the xsd assertion by the way. I didn't write that. 

On October 21, 2015 1:47:44 PM CDT, William Scheidegger <notifications@github.com> wrote:
>> @@ -162,6 +162,10 @@
>>                     xerces:message="Only one default limit group may
>be defined"
>>                     saxon:message="Only one default limit group may
>be defined"/>
>>  
>> +        <xs:assert vc:minVersion="1.1"
>test="count(distinct-values(limit-group/limit/@id)) =
>count(limit-group/limit/@id)"
>> +                   xerces:message="ALL Limits must have unique ids
>across all limit groups"
>> +                   saxon:message="ALL Limits must have unique ids
>across all limit groups" xpathDefaultNamespace="##targetNamespace"/>
>> +
>>          <!-- TODO: assert at least one limit-group or
>global-limit-group is present -->
>
>Easy enough.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/rackerlabs/repose/pull/1398/files#r42665423

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

wdschei commented:
So does branch `REP-2233-fix` supercede branch `REP-2233`?
In other words, is branch `REP-2233` ready to be deleted?

dkowis commented:
Rep2233 has all sorts of debugging in it and it shows where the actual problem is. I'm not sure if that is still desired to see if we want to do something other than just the xsd fix. 

If we don't need that debugging info anymore, then it can be clobbered. 

On October 21, 2015 1:56:39 PM CDT, William Scheidegger <notifications@github.com> wrote:
>So does branch `REP-2233-fix` supercede branch `REP-2233`?
>In other words, is branch `REP-2233` ready to be deleted?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/rackerlabs/repose/pull/1398#issuecomment-149993458

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

wdschei commented:
Cool. I'm going to annotate that branch as such for now.

wdschei commented on repose-aggregator/functional-tests/spock-functional-test/src/test/groovy/features/filters/ratelimiting/CheckRateLimitWConfig.groovy:
```diff
@@ -54,33 +50,16 @@ class CheckRateLimitWConfig extends ReposeValveTest {
-
-*/
 
+    //2233
```
What is this comment for?

dkowis replied:
It used to have an ignore tag indicating it was for bug 2233. I probably had it there so I could find the test easily after closing intellij. 

On October 21, 2015 2:45:07 PM CDT, William Scheidegger <notifications@github.com> wrote:
>>  
>> +    //2233
>
>What is this comment for?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/rackerlabs/repose/pull/1398/files#r42672185

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

wdschei replied:
If this is being called out as the test of the fix for REP-2233, then it should actually say that.

wdschei commented on repose-aggregator/functional-tests/spock-functional-test/src/test/groovy/features/filters/ratelimiting/CheckRateLimitWConfig.groovy:
```diff
@@ -98,8 +77,8 @@ class CheckRateLimitWConfig extends ReposeValveTest {
-        //**defect on get limit for multi group config issue REP-2233
-        //["X-PP-Groups": "user"]             | ["X-PP-User": "default"]     | defaultlimit
+        //This one should also pass, but it's being affected by a state bug: REP-2233
+        ["X-PP-Groups": "user"]             | ["X-PP-User": "default"]     | defaultlimit
```
If this test is now fixed, then shouldn't the comment be removed?

dkowis replied:
Ah yes. It isn't a problem any more. 

On October 21, 2015 2:48:18 PM CDT, William Scheidegger <notifications@github.com> wrote:
>> @@ -98,8 +77,8 @@ class CheckRateLimitWConfig extends ReposeValveTest
>{
>>          ["X-PP-Groups": "multi-limits"]     | ["X-PP-User":
>"multilimits"] | multilimit
>>          ["X-PP-Groups": "query-limits"]     | ["X-PP-User":
>"querylimits"] | querylimit
>>          ["X-PP-Groups": "unlimited"]        | ["X-PP-User":
>"unlimited"]   | unlimitedlimit
>> -        //**defect on get limit for multi group config issue
>REP-2233
>> -        //["X-PP-Groups": "user"]             | ["X-PP-User":
>"default"]     | defaultlimit
>> +        //This one should also pass, but it's being affected by a
>state bug: REP-2233
>> +        ["X-PP-Groups": "user"]             | ["X-PP-User":
>"default"]     | defaultlimit
>
>If this test is now fixed, then shouldn't the comment be removed?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/rackerlabs/repose/pull/1398/files#r42672560

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

wdschei commented on repose-aggregator/functional-tests/spock-functional-test/src/test/groovy/features/filters/ratelimiting/RateLimitingConflictingIDsMultiFilterTest.groovy:
```diff
@@ -0,0 +1,112 @@
+            deproxy.shutdown()
+    }
+
+    //2233
```
If this is being called out as the fix for REP-2233, then it should actually say that.

wdschei commented on repose-aggregator/functional-tests/spock-functional-test/src/test/groovy/features/filters/ratelimiting/RatelimitingMultiUserStateTest.groovy:
```diff
@@ -0,0 +1,169 @@
+            deproxy.shutdown()
+    }
+
+    //2233
```
If this is being called out as the test of the fix for REP-2233, then it should actually say that.

wdschei commented:
The fix looks good as does all the updated/added tests.
Waiting on Jenkins and QE before calling it good.

jennyvo commented:
yes, it's weird --- fixed that still failed on jenkin

jennyvo commented:
Retest this please.

jennyvo commented:
Retest this please.

jennyvo commented:
Retest this please.

dkowis commented:
Holy crap. This execution made a one gig log file. 

I think there are rate limiting configs that don't validate somewhere in the test code. 

On October 21, 2015 11:29:24 PM CDT, Stabby the Narwhal <notifications@github.com> wrote:
>
>Refer to this link for build results (access rights to CI server
>needed): 
>https://jenkins.openrepose.org//job/repose-pull_requests/517/<h2>Build
>result: FAILURE</span></h2>[...truncated 6762651 lines...][JENKINS]
>Archiving /var/lib/jenkins/jobs/repose-pull_requests/workspace/pom.xml
>to
>org.openrepose/repose/7.2.0.1-SNAPSHOT/repose-7.2.0.1-SNAPSHOT.pom[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/core/core-service-api/pom.xml
>to
>org.openrepose/core-service-api/7.2.0.1-SNAPSHOT/core-service-api-7.2.0.1-SNAPSHOT.pom[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/core/core-service-api/target/core-service-api-7.2.0.1-SNAPSHOT.jar
>to
>org.openrepose/core-service-api/7.2.0.1-SNAPSHOT/core-service-api-7.2.0.1-SNAPSHOT.jar[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/core/pom.xml
>to
>org.openrepose/core-support/7.2.0.1-SNAPSHOT/core-support-7.2.0.1-SNAPSHOT.pom[JENKINS]
>Archiving /var/lib/jenkins/jobs/repose-pull_requests/worksp
> ace/repo
>se-aggregator/components/filters/uri-stripper/pom.xml to
>org.openrepose/uri-stripper/7.2.0.1-SNAPSHOT/uri-stripper-7.2.0.1-SNAPSHOT.pom[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/components/filters/uri-stripper/target/uri-stripper-7.2.0.1-SNAPSHOT.jar
>to
>org.openrepose/uri-stripper/7.2.0.1-SNAPSHOT/uri-stripper-7.2.0.1-SNAPSHOT.jar[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/components/filters/echo/pom.xml
>to
>org.openrepose/echo/7.2.0.1-SNAPSHOT/echo-7.2.0.1-SNAPSHOT.pom[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/components/filters/echo/target/echo-7.2.0.1-SNAPSHOT.jar
>to
>org.openrepose/echo/7.2.0.1-SNAPSHOT/echo-7.2.0.1-SNAPSHOT.jar[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/functional-tests/spock-functional-test/pom.xml
>to org.openrepose/repose-spock-tests/7.2.0.1-SNAPSHOT/repose-spock-tes
> ts-7.2.0
>.1-SNAPSHOT.pom[JENKINS] Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/functional-tests/spock-functional-test/target/repose-spock-tests-7.2.0.1-SNAPSHOT.jar
>to
>org.openrepose/repose-spock-tests/7.2.0.1-SNAPSHOT/repose-spock-tests-7.2.0.1-SNAPSHOT.jar[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/installation/deb/pom.xml
>to
>org.openrepose/deb/7.2.0.1-SNAPSHOT/deb-7.2.0.1-SNAPSHOT.pom[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/components/filters/header-identity/pom.xml
>to
>org.openrepose/header-identity/7.2.0.1-SNAPSHOT/header-identity-7.2.0.1-SNAPSHOT.pom[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/components/filters/header-identity/target/header-identity-7.2.0.1-SNAPSHOT.jar
>to
>org.openrepose/header-identity/7.2.0.1-SNAPSHOT/header-identity-7.2.0.1-SNAPSHOT.jar[JENKINS]
>Archiving /var/lib/jenkins/jobs/repos
> e-pull_r
>equests/workspace/repose-aggregator/installation/rpm/repose-filter-bundle/pom.xml
>to
>org.openrepose/repose-rpm-filter-bundle/7.2.0.1-SNAPSHOT/repose-rpm-filter-bundle-7.2.0.1-SNAPSHOT.pom[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/components/filters/translation/pom.xml
>to
>org.openrepose/translation/7.2.0.1-SNAPSHOT/translation-7.2.0.1-SNAPSHOT.pom[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/components/filters/translation/target/translation-7.2.0.1-SNAPSHOT.jar
>to
>org.openrepose/translation/7.2.0.1-SNAPSHOT/translation-7.2.0.1-SNAPSHOT.jar[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/experimental/exception-filter/pom.xml
>to
>org.openrepose/exception-filter/7.2.0.1-SNAPSHOT/exception-filter-7.2.0.1-SNAPSHOT.pom[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/experimental/exception-filter/target/except
> ion-filt
>er-7.2.0.1-SNAPSHOT.jar to
>org.openrepose/exception-filter/7.2.0.1-SNAPSHOT/exception-filter-7.2.0.1-SNAPSHOT.jar[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/services/datastore/impl/pom.xml
>to
>org.openrepose/datastore-service/7.2.0.1-SNAPSHOT/datastore-service-7.2.0.1-SNAPSHOT.pom[JENKINS]
>Archiving
>/var/lib/jenkins/jobs/repose-pull_requests/workspace/repose-aggregator/services/datastore/impl/target/datastore-service-7.2.0.1-SNAPSHOT.jar
>to
>org.openrepose/datastore-service/7.2.0.1-SNAPSHOT/datastore-service-7.2.0.1-SNAPSHOT.jarchannel
>stoppedIRC notifier plugin: Sending notification to: #reposeNotifying
>upstream projects of job completionSetting status of
>8dc972a8a56d4f39a26b2482639c5fa48c1692c3 to FAILURE with url
>http://jenkins.openrepose.org/job/repose-pull_requests/517/ and
>message: 'Build finished. No test results found.'
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/rackerlabs/repose/pull/1398#issuecomment-150103662

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

jennyvo commented:
Retest this please.

jennyvo commented:
don't know why 1 test keeps time out.

dkowis commented:
Run it locally? 

On October 22, 2015 1:05:56 PM CDT, Jenny Vo <notifications@github.com> wrote:
>don't know why 1 test keeps time out.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/rackerlabs/repose/pull/1398#issuecomment-150308178

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

jennyvo commented:
Yes, it ran locally...will try to run the whole suite to see if other previous test cause the jam.

jennyvo commented:
ahhh...rate-limiting config again...thought it's fixed maybe forgot to commit.

dkowis commented:
`2015-10-22 16:00:14,491 14758 [Event Kernel Thread] ERROR org.openrepose.commons.config.parser.jaxb.JaxbConfigurationParser - Failed to utilize the UnmarshallerValidator. Reason: cvc-assertion-failure-mesg: Assertion failed for schema type 'RateLimitingConfiguration'. ALL Limits must have unique ids across all limit groups. `

I'll spend some time to go through the configs in the test code and see if I can get those unique IDs gone. Since we have actual test failures now, it should be easier.

jennyvo commented:
I thought I went through and fixed all unique id issues

jennyvo commented:
something wrong, commit mess up

jennyvo commented:
god...still have many more need to update.

jennyvo commented:
BeepDog, I fixed those don't worry, I hope that's all.

dkowis commented:
Thanks a bunch!

dkowis commented:
I addressed the comments in the pull request, but one of the tests is failing, not because of rate limiting configs, but because of how rate limiting tests generally have state. Something's getting rate limited... Also there's a test in that class `def "Check limit group"` which is annotated with //issue REP-2233. I didn't touch it, becuase it's after midnight now, and sleep beckons.

jennyvo commented:
'Check Limit Group' still return negative number for remaining 
other test said have to honor what ever response code from origin service even if with rate limit.

dkowis commented:
Wat. 

``` 
Notifying upstream projects of job completion Setting status of 14d1471e809bd58ae55383a54ff5bda1bf23f810 to SUCCESS with url http://jenkins.openrepose.org/job/repose-pull_requests/530/ and message: 'Build finished. No test results found.' Refer to this link for build results (access rights to CI server needed): https://jenkins.openrepose.org//job/repose-pull_requests/530/ Finished: SUCCESS 
```

On October 23, 2015 11:12:30 PM CDT, Stabby the Narwhal <notifications@github.com> wrote:
>
>Refer to this link for build results (access rights to CI server
>needed): 
>https://jenkins.openrepose.org//job/repose-pull_requests/530/
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/rackerlabs/repose/pull/1398#issuecomment-150751555

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

dkowis commented:
Oh nevermind. I should be asleep.... 

On October 23, 2015 11:12:30 PM CDT, Stabby the Narwhal <notifications@github.com> wrote:
>
>Refer to this link for build results (access rights to CI server
>needed): 
>https://jenkins.openrepose.org//job/repose-pull_requests/530/
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/rackerlabs/repose/pull/1398#issuecomment-150751555

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

wdschei commented:
Retest this please.

dkowis commented:
Those failures won't be fixed. I thought I pushed changes to fix them, but maybe I only dreamt about it...

dkowis commented:
Forgot to push my code. I am le fail

wdschei commented:
Cool. The test updates look good.
